juju/model.py

```diff
@@ -0,0 +1,166 @@
+            value = [Relation(relation_id) for relation_id in yaml.safe_load(output.decode('utf8'))]
+            self[relation_name] = value
+        return value
+
+
+class Relation:
+    def __init__(self, relation_id, data=None):
+        self.relation_id = relation_id
+        self.data = RelationData(relation_id, data)
+        if data is not None:
+            self._apps = [key for key in self.data if isinstance(key, Application)]
+        else:
+            self._apps = None
+
+    @property
+    def apps(self):
+        if self._apps is None:
+            # Lazy-load the apps on this relation.
+            output = run(['relation-list', '-r', self.relation_id, '--format=yaml'], stdout=PIPE, check=True).stdout
+            member_names = yaml.safe_load(output.decode('utf8'))
+            units_by_app = {}
+            for member_name in sorted(member_names):
+                if '/' not in member_name:
+                    # Handle possible future support from Juju for application relation data.
+                    continue
+                app_name = member_name.split('/')[0]
+                units_by_app.setdefault(app_name, []).append(Unit(member_name))
+            self._apps = [Application(app_name, units) for app_name, units in sorted(units_by_app.items())]
+        return self._apps
+
+
+class RelationData(dict):
+    def __init__(self, relation_id, relation_data=None):
+        super().__init__()
+        self.relation_id = relation_id
+        if relation_data is not None:
+            app_units = {}
+            for app_or_unit_name, member_data in sorted(relation_data.items()):
+                # Apps and units are represented in the same data structure to match the expected
+                # pattern for interacting with Juju and how we present it to the charm code. However,
+                # this means we have to distinguish them by the presence or lack of a / in the name.
+                if '/' not in app_or_unit_name:
+                    # Skip the apps until we have all of the info about which
+                    # of its units are available on the relation.
+                    continue
+                unit = Unit(app_or_unit_name)
+                app_name = unit.name.split('/')[0]
+                app_units.setdefault(app_name, []).append(unit)
+                self[unit] = member_data
+            for app_name, units in app_units.items():
+                app = Application(app_name, units)
+                self[app] = relation_data[app.name]
+
+    def __getitem__(self, app_or_unit):
+        if app_or_unit not in self:
+            # Lazy-load the data for this relation member.
+            # TODO: We should also lazy-load the list of members so that
+            # KeyErrors can be raised if invalid members are accessed.
+            output = run(['relation-get', '-r', self.relation_id, '-', app_or_unit.name, '--format=yaml'], stdout=PIPE, check=True).stdout
```

Comment:
Out of curiosity: does it support other formats other than YAML nowadays? Something more machine oriented, rather than human oriented, that still preserves all typing in the original data (JSON, maybe)?

Reply:
```
--format  (= smart)
    Specify output format (json|smart|yaml)
```

I can switch all of these to JSON and use the stdlib JSON library as well.